In [1]:
import sys
assert sys.version_info >= (3, 7)
from packaging import version
import sklearn
assert version.parse(sklearn.__version__) >= version.parse("1.0.1")
import pandas as pd
import numpy as np 
import seaborn as sns                 
import matplotlib.pyplot as plt             
#%matplotlib inline     
sns.set(color_codes=True)
from IPython.display import display
import matplotlib.pyplot as plt
from sklearn import tree
import matplotlib.pyplot as plt
from scipy import stats
import numpy as np
from matplotlib.ticker import MaxNLocator
from matplotlib.ticker import PercentFormatter

In [2]:
df=pd.read_csv("C:/Users/mehri/Downloads/BankChurners.csv")

## Data Exploration & Preparation

In [3]:
df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

In [4]:
df.info

<bound method DataFrame.info of        CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0      768805383  Existing Customer            45      M                3   
1      818770008  Existing Customer            49      F                5   
2      713982108  Existing Customer            51      M                3   
3      769911858  Existing Customer            40      F                4   
4      709106358  Existing Customer            40      M                3   
...          ...                ...           ...    ...              ...   
10122  772366833  Existing Customer            50      M                2   
10123  710638233  Attrited Customer            41      M                2   
10124  716506083  Attrited Customer            44      F                1   
10125  717406983  Attrited Customer            30      M                2   
10126  714337233  Attrited Customer            43      F                2   

      Education_Level Marital_Status Income

In [5]:
#data type
df.dtypes

CLIENTNUM                                                                                                                               int64
Attrition_Flag                                                                                                                         object
Customer_Age                                                                                                                            int64
Gender                                                                                                                                 object
Dependent_count                                                                                                                         int64
Education_Level                                                                                                                        object
Marital_Status                                                                                                                         object
Income

In [6]:
df.describe()

CLIENTNUM  Customer_Age  Dependent_count  Months_on_book  \
count  1.012700e+04  10127.000000     10127.000000    10127.000000   
mean   7.391776e+08     46.325960         2.346203       35.928409   
std    3.690378e+07      8.016814         1.298908        7.986416   
min    7.080821e+08     26.000000         0.000000       13.000000   
25%    7.130368e+08     41.000000         1.000000       31.000000   
50%    7.179264e+08     46.000000         2.000000       36.000000   
75%    7.731435e+08     52.000000         3.000000       40.000000   
max    8.283431e+08     73.000000         5.000000       56.000000   

       Total_Relationship_Count  Months_Inactive_12_mon  \
count              10127.000000            10127.000000   
mean                   3.812580                2.341167   
std                    1.554408                1.010622   
min                    1.000000                0.000000   
25%                    3.000000                2.000000   
50%                    4.000000                2.000000   
75%                    5.000000                3.000000   
max                    6.000000                6.000000   

       Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
count           10127.000000  10127.000000         10127.000000   
mean                2.455317   8631.953698          1162.814061   
std                 1.106225   9088.776650           814.987335   
min                 0.000000   1438.300000             0.000000   
25%                 2.000000   2555.000000           359.000000   
50%                 2.000000   4549.000000          1276.000000   
75%                 3.000000  11067.500000          1784.000000   
max                 6.000000  34516.000000          2517.000000   

       Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
count     10127.000000          10127.000000     10127.000000    10127.000000   
mean       7469.139637              0.759941      4404.086304       64.858695   
std        9090.685324              0.219207      3397.129254       23.472570   
min           3.000000              0.000000       510.000000       10.000000   
25%        1324.500000              0.631000      2155.500000       45.000000   
50%        3474.000000              0.736000      3899.000000       67.000000   
75%        9859.000000              0.859000      4741.000000       81.000000   
max       34516.000000              3.397000     18484.000000      139.000000   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
count         10127.000000           10127.000000   
mean              0.712222               0.274894   
std               0.238086               0.275691   
min               0.000000               0.000000   
25%               0.582000               0.023000   
50%               0.702000               0.176000   
75%               0.818000               0.503000   
max               3.714000               0.999000   

       Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
count                                       10127.000000                                                                                    
mean                                            0.159997                                                                                    
std                                             0.365301                                                                                    
min                                             0.000008                                                                                    
25%                                             0.000099                                                                                    
50%                                             0.000181                                                                                    
75%                                             

In [7]:
df.shape

(10127, 23)

In [8]:
df.nunique()

CLIENTNUM                                                                                                                             10127
Attrition_Flag                                                                                                                            2
Customer_Age                                                                                                                             45
Gender                                                                                                                                    2
Dependent_count                                                                                                                           6
Education_Level                                                                                                                           7
Marital_Status                                                                                                                            4
Income_Category     

In [9]:
df.isnull().sum()

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

## Data Preprocessing

### One-Hot Encoding of Categorical Variables

One-hot encoding is performed on the categorical variables such as 'Attrition_Flag', 'Gender', 'Education_Level', 'Marital_Status', 'Income_Category', and 'Card_Category' to convert them into a format that can be provided to machine learning algorithms to enable better model training. This process transforms categorical variables into a set of binary variables (0s and 1s), with each binary variable representing a unique category, which is necessary since most machine learning models require numerical input.

In [10]:
# Perform one-hot encoding on the specified categorical variables
categorical_columns = ['Attrition_Flag', 'Gender', 'Education_Level', 'Marital_Status', 'Income_Category', 'Card_Category']
df_encoded = pd.get_dummies(df, columns=categorical_columns)

# Display the shape of the new dataframe and the first few rows to confirm the encoding
df_encoded_shape = df_encoded.shape
df_encoded_head = df_encoded.head()

df_encoded_shape, df_encoded_head

((10127, 42),
    CLIENTNUM  Customer_Age  Dependent_count  Months_on_book  \
 0  768805383            45                3              39   
 1  818770008            49                5              44   
 2  713982108            51                3              36   
 3  769911858            40                4              34   
 4  709106358            40                3              21   
 
    Total_Relationship_Count  Months_Inactive_12_mon  Contacts_Count_12_mon  \
 0                         5                       1                      3   
 1                         6                       1                      2   
 2                         4                       1                      0   
 3                         3                       4                      1   
 4                         5                       1                      0   
 
    Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  ...  \
 0       12691.0                  777          11914.0  ...  

### Unique Identifier

The 'CLIENTNUM' column has been removed from the dataset as it serves as a unique identifier for each customer, which does not provide any predictive power for the model and could potentially skew the learning process if included.

In [11]:
# Remove the 'CLIENTNUM' column as it is a unique identifier and not useful for model learning
df_encoded_dropped = df_encoded.drop('CLIENTNUM', axis=1)

# Display the shape of the new dataframe and the first few rows to confirm the column has been removed
df_encoded_dropped_shape = df_encoded_dropped.shape
df_encoded_dropped_head = df_encoded_dropped.head()

df_encoded_dropped_shape, df_encoded_dropped_head

((10127, 41),
    Customer_Age  Dependent_count  Months_on_book  Total_Relationship_Count  \
 0            45                3              39                         5   
 1            49                5              44                         6   
 2            51                3              36                         4   
 3            40                4              34                         3   
 4            40                3              21                         5   
 
    Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
 0                       1                      3       12691.0   
 1                       1                      2        8256.0   
 2                       1                      0        3418.0   
 3                       4                      1        3313.0   
 4                       1                      0        4716.0   
 
    Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  ...  \
 0                  777          11

### Missing Values

The dataset was scanned for columns containing the placeholder "Unknown", which were treated as missing values since these do not contribute meaningful information for model training. These "Unknown" values have been replaced with NaN to standardize the representation of missing data, facilitating the application of imputation strategies in subsequent preprocessing steps. This ensures that our model learns from genuine data distributions without bias from arbitrary placeholders.

In [12]:
# Check for columns that contain the placeholder "Unknown" to treat them as missing values
unknown_counts = df_encoded_dropped.apply(lambda x: (x == 'Unknown').sum())
unknown_columns = unknown_counts[unknown_counts > 0].index.tolist()

# For demonstration, let's replace "Unknown" with NaN (to treat them as missing values) and then decide on an imputation strategy
df_encoded_dropped.replace('Unknown', pd.NA, inplace=True)

# Check the counts of NaN values in the dataset after replacing "Unknown" with NaN
nan_counts_after_replacement = df_encoded_dropped.isna().sum()

unknown_columns, nan_counts_after_replacement[nan_counts_after_replacement > 0]

([], Series([], dtype: int64))

### Feature Scaling

Feature scaling has been applied to the numerical columns using `StandardScaler` to normalize the distribution of the features. This is a crucial step before model training because it ensures that all numerical features contribute equally to the model's performance, avoiding any bias that might occur due to the scale of the features. Scaled features have a mean of 0 and a standard deviation of 1, which is particularly important for models that are sensitive to the magnitude of inputs, such as SVM and k-NN.

In [13]:
from sklearn.preprocessing import StandardScaler

# List of numerical features to scale
numerical_features = [
    'Customer_Age', 'Dependent_count', 'Months_on_book',
    'Total_Relationship_Count', 'Months_Inactive_12_mon', 'Contacts_Count_12_mon',
    'Credit_Limit', 'Total_Revolving_Bal', 'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1', 'Total_Trans_Amt', 'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1', 'Avg_Utilization_Ratio'
]

# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler to the numerical features and transform them
df_encoded_dropped[numerical_features] = scaler.fit_transform(df_encoded_dropped[numerical_features])

# Display the first few rows of the dataset to confirm the scaling
df_encoded_dropped.head()

Customer_Age  Dependent_count  Months_on_book  Total_Relationship_Count  \
0     -0.165406         0.503368        0.384621                  0.763943   
1      0.333570         2.043199        1.010715                  1.407306   
2      0.583058         0.503368        0.008965                  0.120579   
3     -0.789126         1.273283       -0.241473                 -0.522785   
4     -0.789126         0.503368       -1.869317                  0.763943   

   Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
0               -1.327136               0.492404      0.446622   
1               -1.327136              -0.411616     -0.041367   
2               -1.327136              -2.219655     -0.573698   
3                1.641478              -1.315636     -0.585251   
4               -1.327136              -2.219655     -0.430877   

   Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  ...  \
0            -0.473422         0.488971              2.623494  ...   
1            -0.366667        -0.008486              3.563293  ...   
2            -1.426858        -0.445658              8.367214  ...   
3             1.661686        -0.734100              2.942843  ...   
4            -1.426858        -0.302868              6.455682  ...   

   Income_Category_$120K +  Income_Category_$40K - $60K  \
0                        0                            0   
1                        0                            0   
2                        0                            0   
3                        0                            0   
4                        0                            0   

   Income_Category_$60K - $80K  Income_Category_$80K - $120K  \
0                            1                             0   
1                            0                             0   
2                            0                             1   
3                            0                             0   
4                            1                             0   

   Income_Category_Less than $40K  Income_Category_Unknown  \
0                               0                        0   
1                               1                        0   
2                               0                        0   
3                               1                        0   
4                               0                        0   

   Card_Category_Blue  Card_Category_Gold  Card_Category_Platinum  \
0                   1                   0                       0   
1                   1                   0                       0   
2                   1                   0                       0   
3                   1                   0                       0   
4                   1                   0                       0   

   Card_Category_Silver  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 41 columns]

### Irrelevant Features

Columns that are likely outputs from a previous Naive Bayes classifier model were identified and removed from the dataset. These features are considered irrelevant for the development of new model predictions, as they may introduce bias and do not constitute raw input data. Their removal helps ensure that the predictive model we develop is based purely on actual observed data and not influenced by the artifacts of a previous model's output.

In [31]:
# Remove the last two columns which are likely outputs from a Naive Bayes classifier and not useful for new model predictions
columns_to_remove = [
    'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
    'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'
]

df_final = df_encoded_dropped.drop(columns=columns_to_remove)

# Display the shape of the new dataframe and the first few rows to confirm the removal of the irrelevant features
df_final_shape = df_final.shape
df_final_head = df_final.head()

df_final_shape, df_final_head

((10127, 39),
    Customer_Age  Dependent_count  Months_on_book  Total_Relationship_Count  \
 0     -0.165406         0.503368        0.384621                  0.763943   
 1      0.333570         2.043199        1.010715                  1.407306   
 2      0.583058         0.503368        0.008965                  0.120579   
 3     -0.789126         1.273283       -0.241473                 -0.522785   
 4     -0.789126         0.503368       -1.869317                  0.763943   
 
    Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
 0               -1.327136               0.492404      0.446622   
 1               -1.327136              -0.411616     -0.041367   
 2               -1.327136              -2.219655     -0.573698   
 3                1.641478              -1.315636     -0.585251   
 4               -1.327136              -2.219655     -0.430877   
 
    Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  ...  \
 0            -0.473422         0.4

In [33]:
# Drop Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1 and 2
df = df.drop(['Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1', 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'], axis=1)

## Correlation Analysis

The correlation matrix is computed to quantify the linear relationships between numerical features. This step is critical for identifying features that are closely related to each other, which can indicate redundancy. For instance, a high positive correlation between `Credit_Limit` and `Avg_Open_To_Buy` suggests that these features may convey similar information, and one of them could potentially be excluded to simplify the model. Similarly, a strong positive correlation between `Total_Trans_Amt` and `Total_Trans_Ct` indicates that as the number of transactions increases, the total transaction amount tends to increase as well. By identifying such relationships, we can prevent multicollinearity, which can obscure the interpretation of feature importance and reduce model performance.

In [15]:
# Calculate the correlation matrix for the final dataset
correlation_matrix = df_final.corr()

# Display the correlation matrix
correlation_matrix

Customer_Age  Dependent_count  \
Customer_Age                          1.000000        -0.122254   
Dependent_count                      -0.122254         1.000000   
Months_on_book                        0.788912        -0.103062   
Total_Relationship_Count             -0.010931        -0.039076   
Months_Inactive_12_mon                0.054361        -0.010768   
Contacts_Count_12_mon                -0.018452        -0.040505   
Credit_Limit                          0.002476         0.068065   
Total_Revolving_Bal                   0.014780        -0.002688   
Avg_Open_To_Buy                       0.001151         0.068291   
Total_Amt_Chng_Q4_Q1                 -0.062042        -0.035439   
Total_Trans_Amt                      -0.046446         0.025046   
Total_Trans_Ct                       -0.067097         0.049912   
Total_Ct_Chng_Q4_Q1                  -0.012143         0.011087   
Avg_Utilization_Ratio                 0.007114        -0.037135   
Attrition_Flag_Attrited Customer      0.018203         0.018991   
Attrition_Flag_Existing Customer     -0.018203        -0.018991   
Gender_F                              0.017312        -0.004563   
Gender_M                             -0.017312         0.004563   
Education_Level_College              -0.014788         0.003369   
Education_Level_Doctorate             0.025199        -0.003368   
Education_Level_Graduate             -0.000203         0.000671   
Education_Level_High School           0.001199        -0.013127   
Education_Level_Post-Graduate        -0.022081         0.009459   
Education_Level_Uneducated            0.005057         0.002190   
Education_Level_Unknown               0.005377         0.004922   
Marital_Status_Divorced              -0.042614         0.006697   
Marital_Status_Married                0.047364         0.014385   
Marital_Status_Single                -0.011248        -0.040707   
Marital_Status_Unknown               -0.026694         0.041743   
Income_Category_$120K +               0.044332         0.024831   
Income_Category_$40K - $60K          -0.013804        -0.013493   
Income_Category_$60K - $80K          -0.017869         0.028975   
Income_Category_$80K - $120K          0.005381         0.047611   
Income_Category_Less than $40K       -0.002573        -0.046785   
Income_Category_Unknown              -0.002264        -0.019205   
Card_Category_Blue                    0.021409        -0.027671   
Card_Category_Gold                   -0.011901         0.027035   
Card_Category_Platinum                0.006515         0.003555   
Card_Category_Silver                 -0.019425         0.017322   

                                  Months_on_book  Total_Relationship_Count  \
Customer_Age                            0.788912                 -0.010931   
Dependent_count                        -0.103062                 -0.039076   
Months_on_book                          1.000000                 -0.009203   
Total_Relationship_Count               -0.009203                  1.000000   
Months_Inactive_12_mon                  0.074164                 -0.003675   
Contacts_Count_12_mon                  -0.010774                  0.055203   
Credit_Limit                            0.007507                 -0.071386   
Total_Revolving_Bal                     0.008623                  0.013726   
Avg_Open_To_Buy                         0.006732                 -0.072601   
Total_Amt_Chng_Q4_Q1                   -0.048959                  0.050119   
Total_Trans_Amt                        -0.038591                 -0.347229   
Total_Trans_Ct                         -0.049819                 -0.241891   
Total_Ct_Chng_Q4_Q1                    -0.014072                  0.040831   
Avg_Utilization_Ratio                  -0.007541                  0.067663   
Attrition_Flag_Attrited Customer        0.013687                 -0.150005   
Attrition_Flag_Existing Customer       -0.013687                  0.150005   
Gender_F                   

In [24]:
# Performing correlation analysis among numerical features using Pearson correlation

# Calculate the Pearson correlation matrix
correlation_matrix = X_encoded[numerical_features].corr()

# Identify highly correlated pairs (above 0.8 or below -0.8, excluding self-correlation of 1)
highly_correlated_pairs = correlation_matrix.where(np.triu(np.abs(correlation_matrix) > 0.8, k=1)).stack().reset_index()
highly_correlated_pairs.columns = ['Feature1', 'Feature2', 'Correlation']

highly_correlated_pairs

Feature1         Feature2  Correlation
0     Credit_Limit  Avg_Open_To_Buy     0.995981
1  Total_Trans_Amt   Total_Trans_Ct     0.807192

The correlation analysis among numerical features using Pearson correlation has identified the following highly correlated pairs:

* Credit_Limit and Avg_Open_To_Buy with a correlation coefficient of approximately 0.996. This high correlation suggests that these two features carry redundant information, as they move together very closely.

* Total_Trans_Amt (Total Transaction Amount) and Total_Trans_Ct (Total Transaction Count) with a correlation coefficient of approximately 0.807. This indicates a strong positive relationship, suggesting that as the number of transactions increases, the total amount of transactions also increases.

## Feature Encoding

Feature encoding is conducted to transform categorical variables into a machine-readable format. The `Attrition_Flag` column, which indicates customer churn, is encoded as a binary variable, with '1' representing 'Attrited Customer' and '0' for others. This binary encoding is essential for classification models that require numerical input.

Based on the earlier correlation analysis, features such as `Avg_Open_To_Buy` and `Total_Trans_Ct` that showed a high correlation with other features were removed to reduce multicollinearity. This step is taken to ensure that each feature contributes unique information to the model, improving interpretability and potentially enhancing model performance by reducing overfitting.

In [17]:
# Check if 'Attrition_Flag' exists in the dataframe and is categorical
if 'Attrition_Flag' in df.columns and df['Attrition_Flag'].dtype == 'object':
    categorical_cols = df.select_dtypes(include=['object']).columns.drop('Attrition_Flag')
else:
    categorical_cols = df.select_dtypes(include=['object']).columns

# Encode the categorical features
X_encoded = pd.get_dummies(df.drop(columns=['CLIENTNUM', 'Attrition_Flag']), columns=categorical_cols, drop_first=True)
y = df['Attrition_Flag'].apply(lambda x: 1 if x == 'Attrited Customer' else 0)

In [28]:
# Removing the features "Avg_Open_To_Buy" and "Total_Trans_Ct" based on the correlation analysis

X_encoded_reduced = X_encoded.drop(['Avg_Open_To_Buy', 'Total_Trans_Ct'], axis=1)

# Display the first few rows of the reduced feature set to confirm the removal
X_encoded_reduced.head()

Customer_Age  Dependent_count  Months_on_book  Total_Relationship_Count  \
0            45                3              39                         5   
1            49                5              44                         6   
2            51                3              36                         4   
3            40                4              34                         3   
4            40                3              21                         5   

   Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
0                       1                      3       12691.0   
1                       1                      2        8256.0   
2                       1                      0        3418.0   
3                       4                      1        3313.0   
4                       1                      0        4716.0   

   Total_Revolving_Bal  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  ...  \
0                  777                 1.335             1144  ...   
1                  864                 1.541             1291  ...   
2                    0                 2.594             1887  ...   
3                 2517                 1.405             1171  ...   
4                    0                 2.175              816  ...   

   Marital_Status_Single  Marital_Status_Unknown  Income_Category_$40K - $60K  \
0                      0                       0                            0   
1                      1                       0                            0   
2                      0                       0                            0   
3                      0                       1                            0   
4                      0                       0                            0   

   Income_Category_$60K - $80K  Income_Category_$80K - $120K  \
0                            1                             0   
1                            0                             0   
2                            0                             1   
3                            0                             0   
4                            1                             0   

   Income_Category_Less than $40K  Income_Category_Unknown  \
0                               0                        0   
1                               1                        0   
2                               0                        0   
3                               1                        0   
4                               0                        0   

   Card_Category_Gold  Card_Category_Platinum  Card_Category_Silver  
0                   0                       0                     0  
1                   0                       0                     0  
2                   0                       0                     0  
3                   0                       0                     0  
4                   0                       0                     0  

[5 rows x 32 columns]

## Removing Constant & Duplicate Features 

In the data preprocessing phase, constant features were identified and removed. Constant features, which have the same value for all observations, do not contribute any information that can be used for prediction and can be eliminated to reduce model complexity and computational cost.

Additionally, duplicate features were identified and removed. Duplicate features are those that are identical to others across all observations and therefore redundant. Removing duplicates helps in reducing the dimensionality of the dataset, ensuring that each feature provides unique information to the model and improving the efficiency of model training.

In [18]:
# Identifying constant features
constant_features = [col for col in df.columns if df[col].nunique() == 1]

# Dropping constant features
df.drop(columns=constant_features, inplace=True)

# Identifying duplicate features
duplicate_features = df.columns[df.T.duplicated()]

# Dropping duplicate features
df.drop(columns=duplicate_features, inplace=True)

# Display the removed constant and duplicate features
constant_features, duplicate_features


([], Index([], dtype='object'))

## Filter Methods - Statistical Significance Testing

### Mutual Information Method

Mutual Information (MI) has been utilized to quantify the amount of information each feature shares with the target variable, `Attrition_Flag_Attrited Customer`. This method helps in identifying features that have the strongest relationships with the target variable. Features with higher MI scores are more predictive and thus are likely more important for our model. This non-linear dependency evaluation is crucial for understanding the underlying patterns in the data that may not be captured by linear methods such as correlation. The MI scores have been sorted in descending order to prioritize the features that could contribute most to the model's predictive power.

In [21]:
from sklearn.feature_selection import mutual_info_classif

# Prepare the features and target variable for mutual information
X = df_final.drop(['Attrition_Flag_Attrited Customer', 'Attrition_Flag_Existing Customer'], axis=1)
y = df_final['Attrition_Flag_Attrited Customer']

# Calculate Mutual Information
mi_scores = mutual_info_classif(X, y, random_state=0)

# Create a Series for better visualization
mi_scores_series = pd.Series(mi_scores, index=X.columns).sort_values(ascending=False)

mi_scores_series

Total_Trans_Amt                   0.158537
Total_Trans_Ct                    0.110403
Total_Ct_Chng_Q4_Q1               0.091953
Total_Revolving_Bal               0.087675
Avg_Utilization_Ratio             0.044043
Total_Amt_Chng_Q4_Q1              0.029373
Contacts_Count_12_mon             0.025926
Months_Inactive_12_mon            0.024047
Avg_Open_To_Buy                   0.020988
Total_Relationship_Count          0.009936
Credit_Limit                      0.008016
Dependent_count                   0.006807
Education_Level_High School       0.006314
Customer_Age                      0.005603
Card_Category_Platinum            0.005401
Education_Level_Doctorate         0.005297
Income_Category_$60K - $80K       0.004727
Income_Category_$40K - $60K       0.003328
Marital_Status_Single             0.002796
Income_Category_Less than $40K    0.002490
Education_Level_Uneducated        0.001559
Income_Category_$80K - $120K      0.000959
Card_Category_Blue                0.000419
Card_Catego

### Chi-Squared Test

The Chi-Squared Test is utilized here to evaluate the independence of binary features derived from one-hot encoding with respect to the target variable, `Attrition_Flag_Attrited Customer`. Although our dataset is preprocessed with all features encoded and scaled, the Chi-Squared Test is still applicable because it requires categorical input and can be used on binary features. The test provides Chi2 scores and p-values for each feature, with a lower p-value indicating a feature that is likely to be more important and statistically significant with respect to the target variable. Features with higher Chi2 scores and lower p-values are prioritized as they are more likely to have a meaningful association with the target, helping in the selection of relevant features for the predictive model.


In [22]:
from sklearn.feature_selection import chi2

# Since our dataset is already encoded and scaled, we directly select binary features for the Chi-squared test.
# For this example, we'll consider all features as they are encoded and treat them for the Chi-squared test.
# Note: Normally, Chi-squared is applied before encoding and scaling, primarily on raw categorical data.

# Filter for binary features only (resulting from one-hot encoding)
binary_features = [col for col in X.columns if df_final[col].nunique() == 2]

# Select only binary features for this test
X_binary = X[binary_features]

# Apply the Chi-squared test
chi_scores, chi_p_values = chi2(X_binary, y)

# Create a DataFrame for better visualization
chi_results = pd.DataFrame({'Feature': binary_features, 'Chi2 Score': chi_scores, 'P-value': chi_p_values}).sort_values(by='Chi2 Score', ascending=False)

chi_results

Feature  Chi2 Score   P-value
3        Education_Level_Doctorate    8.355706  0.003845
1                         Gender_M    7.443223  0.006368
15     Income_Category_$60K - $80K    6.948616  0.008388
0                         Gender_F    6.624996  0.010056
17  Income_Category_Less than $40K    3.313865  0.068698
10          Marital_Status_Married    3.064753  0.080007
11           Marital_Status_Single    2.241029  0.134392
6    Education_Level_Post-Graduate    1.190018  0.275326
21          Card_Category_Platinum    1.183806  0.276582
14     Income_Category_$40K - $60K    1.138962  0.285872
5      Education_Level_High School    1.116349  0.290706
13         Income_Category_$120K +    0.863454  0.352774
12          Marital_Status_Unknown    0.743541  0.388529
8          Education_Level_Unknown    0.698099  0.403423
22            Card_Category_Silver    0.686162  0.407473
4         Education_Level_Graduate    0.572763  0.449163
2          Education_Level_College    0.560251  0.454159
18         Income_Category_Unknown    0.464578  0.495492
20              Card_Category_Gold    0.357110  0.550115
16    Income_Category_$80K - $120K    0.102784  0.748515
7       Education_Level_Uneducated    0.018020  0.893215
19              Card_Category_Blue    0.007148  0.932623
9          Marital_Status_Divorced    0.006774  0.934405

## Feature Selection

Feature selection has been performed using three different methods to identify the most informative features for our predictive model:

1. **Random Forest Importance:** This method leverages the feature importance attribute provided by Random Forest models. By evaluating the contribution of each feature to the model's performance, we can rank the features based on their importance and select the top features for further analysis.

2. **Recursive Feature Elimination (RFE):** RFE is an iterative feature selection technique that recursively removes features from the dataset and evaluates the model's performance after each removal. By selecting features that contribute the most to the model's accuracy, RFE helps in identifying the subset of features that are most relevant for prediction.

3. **XGBoost Feature Importance:** XGBoost is an ensemble learning algorithm known for its effectiveness in feature selection. Similar to Random Forest, XGBoost provides a feature importance score based on the contribution of each feature to the model's performance. By ranking the features according to their importance, we can identify the key features that have the most significant impact on the predictive power of the model.


## Random Forest

In [36]:
# Re-import necessary libraries and load the dataset, then proceed with feature importance analysis
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.inspection import permutation_importance

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.3, random_state=42)

# Initialize the RandomForest classifier
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the classifier to the training data
rf_classifier.fit(X_train, y_train)

# Obtain feature importance scores from the fitted RandomForest model
feature_importances_rf = rf_classifier.feature_importances_

# Create a DataFrame to display feature importances
feature_importances_df = pd.DataFrame({
    'Feature': X_encoded.columns,
    'Importance': feature_importances_rf
}).sort_values(by='Importance', ascending=False).reset_index(drop=True)

# Perform permutation importance on the test set to get a model-agnostic view of feature importance
perm_importance_result = permutation_importance(rf_classifier, X_test, y_test, n_repeats=10, random_state=42)

# Create a DataFrame to display permutation importance results
perm_importance_df = pd.DataFrame({
    'Feature': X_encoded.columns,
    'Importance': perm_importance_result.importances_mean
}).sort_values(by='Importance', ascending=False).reset_index(drop=True)

# Display the top features based on RandomForest feature importance
feature_importances_df.head(), perm_importance_df.head()

(                 Feature  Importance
 0        Total_Trans_Amt    0.186041
 1         Total_Trans_Ct    0.153520
 2    Total_Ct_Chng_Q4_Q1    0.099389
 3    Total_Revolving_Bal    0.096376
 4  Avg_Utilization_Ratio    0.067332,
                     Feature  Importance
 0            Total_Trans_Ct    0.088516
 1           Total_Trans_Amt    0.041987
 2       Total_Revolving_Bal    0.021158
 3  Total_Relationship_Count    0.021125
 4       Total_Ct_Chng_Q4_Q1    0.012702)

## Recursive Feature Elimination (RFE)

In [35]:
from sklearn.feature_selection import RFE

# Assuming X_encoded_reduced and y are already defined

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded_reduced, y, test_size=0.3, random_state=42)

# Initialize the model to be used for feature selection
model = RandomForestClassifier(n_estimators=100, random_state=42)

# Initialize RFE with the model and desired number of features
rfe = RFE(estimator=model, n_features_to_select=10)

# Fit RFE
rfe.fit(X_train, y_train)

# Transform the data according to the RFE
X_train_rfe = rfe.transform(X_train)
X_test_rfe = rfe.transform(X_test)

# Fit the model on the transformed data
model.fit(X_train_rfe, y_train)

# Evaluate the model
print("Model score with selected features:", model.score(X_test_rfe, y_test))

# List of selected features
selected_features = X_encoded_reduced.columns[rfe.support_]
print("Selected features:", selected_features)

Model score with selected features: 1.0
Selected features: Index(['Total_Relationship_Count', 'Months_Inactive_12_mon',
       'Contacts_Count_12_mon', 'Total_Revolving_Bal', 'Total_Amt_Chng_Q4_Q1',
       'Total_Trans_Amt', 'Total_Ct_Chng_Q4_Q1', 'Avg_Utilization_Ratio',
       'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
       'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'],
      dtype='object')


## XGBoost

In [34]:
import xgboost as xgb
from xgboost import XGBClassifier

# Assuming df is your preprocessed DataFrame and 'y' is the target variable

# Encode categorical variables using get_dummies or another encoding method if not already done
X_encoded = pd.get_dummies(df.drop(['Attrition_Flag'], axis=1))

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.3, random_state=42)

# Initialize the XGBoost classifier
xgb_classifier = XGBClassifier()

# Fit the classifier to the training data
xgb_classifier.fit(X_train, y_train)

# Obtain feature importances from the fitted XGBoost model
feature_importances_xgb = xgb_classifier.feature_importances_

# Create a DataFrame to display feature importances
feature_importances_xgb_df = pd.DataFrame({
    'Feature': X_train.columns,
    'XGBoost Importance': feature_importances_xgb
}).sort_values(by='XGBoost Importance', ascending=False).reset_index(drop=True)

# Display the DataFrame
print(feature_importances_xgb_df)

                           Feature  XGBoost Importance
0                   Total_Trans_Ct            0.230968
1              Total_Revolving_Bal            0.172559
2         Total_Relationship_Count            0.106204
3                  Total_Trans_Amt            0.053931
4              Total_Ct_Chng_Q4_Q1            0.045088
5           Months_Inactive_12_mon            0.039066
6                     Customer_Age            0.035142
7            Contacts_Count_12_mon            0.030268
8             Total_Amt_Chng_Q4_Q1            0.028360
9                     Credit_Limit            0.021270
10          Marital_Status_Married            0.018849
11                        Gender_F            0.017199
12                 Avg_Open_To_Buy            0.016241
13      Education_Level_Uneducated            0.015431
14          Marital_Status_Unknown            0.015157
15     Income_Category_$60K - $80K            0.014021
16           Avg_Utilization_Ratio            0.012489
17     Inc

## Dimensionality Reduction

As part of the preprocessing pipeline, Principal Component Analysis (PCA) has been applied for dimensionality reduction. PCA is a widely used technique for reducing the number of features in high-dimensional datasets while preserving most of the variance in the data.

The reduced-dimensional representation obtained through PCA can help alleviate issues associated with high-dimensional data, such as the curse of dimensionality, computational complexity, and overfitting. By retaining the most important information while discarding redundant or less informative features, PCA facilitates more efficient model training and interpretation.

In this notebook, PCA has been employed to reduce the dimensionality of the feature space, with the number of principal components chosen based on the desired level of variance retention or reduction in dimensionality. This preprocessing step aims to streamline the modeling process and improve the performance and interpretability of the predictive model.

## Principal Component Analysis (PCA)

In [28]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

# Assuming X_encoded_reduced is our DataFrame after encoding and removing correlated features
# And that y is the target variable

# Standardize the features before applying PCA
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_encoded_reduced)

# Initialize PCA, here we choose enough components to explain 95% of the variance
pca = PCA(n_components=0.95)
X_pca = pca.fit_transform(X_scaled)

# Create a DataFrame of the PCA results
X_pca_df = pd.DataFrame(X_pca, columns=[f'PC{i+1}' for i in range(X_pca.shape[1])])

# Display the shape of the original data and the PCA transformed data
print(f'Original shape: {X_encoded_reduced.shape}')
print(f'Transformed shape: {X_pca_df.shape}')

# Display the DataFrame with PCA components
X_pca_df.head()

Original shape: (10127, 30)
Transformed shape: (10127, 23)


PC1       PC2       PC3       PC4       PC5       PC6       PC7  \
0  1.489641 -0.884369  2.822447  1.027412 -0.692476  2.291951  2.966847   
1 -1.231741  0.990800  1.748798  4.956642  1.129191  0.943643  9.116472   
2  1.367096 -0.644059  4.210599  3.433676  1.566070  0.933379  8.362635   
3 -2.350698  1.368173  2.822851  4.264532  0.942465  2.086084  2.982933   
4  1.178185  1.237027  5.045527  2.527750  0.335862  3.285261  7.355617   

        PC8       PC9      PC10  ...      PC14      PC15      PC16      PC17  \
0 -0.644860 -2.648468  1.549872  ...  0.005604  0.410892 -0.401363  0.038723   
1 -1.409817 -2.851111 -0.311319  ...  0.525862 -1.052708 -0.127305  0.058734   
2 -0.820686 -2.729550 -0.491426  ...  0.565621 -2.350905  0.507948 -0.447944   
3 -1.305153 -0.642003  1.467631  ... -0.167906 -0.533881 -0.017730  0.522513   
4 -0.895749 -3.750263 -1.424266  ... -0.007219 -0.982136 -0.175146 -0.511793   

       PC18      PC19      PC20      PC21      PC22      PC23  
0 -0.010850  0.880408  0.886783 -1.318672 -0.113320  1.114157  
1 -1.259141  0.589209  3.004734 -2.426894 -2.797596  5.443879  
2 -0.133761  0.408147  0.378961 -0.625645  2.097320  0.045741  
3  0.076580 -2.250831  1.537397 -0.383859  0.134424  3.081507  
4 -0.554103  0.677288  0.879807  0.142730  0.882612  1.522363  

[5 rows x 23 columns]

## Choice of Feature Selection Method - Iteration 1*

For the initial iteration of feature selection, a combination of filter and wrapper methods has been employed. Filter methods, such as correlation analysis and chi-squared test, were used to identify features with strong relationships to the target variable or significant predictive power. Additionally, wrapper methods like Recursive Feature Elimination (RFE) were utilized to recursively select the most relevant features based on their contribution to model performance. This approach aims to capture both the intrinsic properties of the features and their interaction with the model, providing a comprehensive selection of informative features for the predictive task.

The most optimal feature selection method involves selecting a subset of features that maximize predictive performance and enhance model generalization.

## Iteration & Refinement

### Step-by-Step Iteration Process:

#### 1. Review Model Performance
- Assess model performance using the metrics from cross-validation.
- Identify areas of weakness, such as high variance (overfitting) or high bias (underfitting).

#### 2. Feature Re-evaluation
- Revisit the feature selection process.
- Consider whether some important features might have been overlooked or if irrelevant features were included.

#### 3. Incorporate Domain Knowledge
- Consult with domain experts to identify features that are known to be important predictors.
- Adjust feature selection based on their input.

#### 4. Feature Engineering
- Create new features based on existing data that could capture more complexity within the dataset.
- Consider polynomial features, interactions between features, or aggregations.

##### 5. Alternative Feature Selection Techniques
- If initial feature selection was done using a filter method, try a wrapper or embedded method, or vice versa.
- Explore algorithms that inherently perform feature selection, such as decision trees or L1 regularized models.

#### 6. Hyperparameter Tuning
- Use grid search or random search to find the optimal set of hyperparameters for your model.
- Consider automated hyperparameter optimization tools like Hyperopt or Optuna.

#### 7. Cross-Validation with New Features
- Perform cross-validation again with the revised set of features or model parameters.
- Ensure that the model is evaluated using the same metrics for consistency.

#### 8. Model Complexity
- If the model is overfitting, consider simplifying it by reducing features or using regularization techniques.
- If the model is underfitting, try more complex models or add features.

#### 9. Ensemble Methods
- Explore ensemble methods that combine predictions from several models to improve overall performance.
- Consider stacking, bagging, or boosting techniques.

#### 10. Validation with an Independent Dataset
- If available, test the model on an independent validation set to assess its generalizability.

#### 11. Iterative Refinement
- Repeat the process, refining features and model choices based on performance on the validation set.

#### 12. Documentation
- Document each iteration's results, including the rationale for changes and their impact.
- Keep track of performance metrics to compare different iterations.


## Alternative Feature Selection Methods



#### 1. Regularization Methods
- **Lasso (L1 Regularization):** Shrinks coefficients to zero, effectively performing feature selection.
- **Ridge (L2 Regularization):** Encourages smaller, more diffuse coefficients.
- **Elastic Net:** Combines L1 and L2 regularization to select features.

#### 2. Model-Based Selection
- **Random Forest Importance:** Leverages the feature importance from Random Forests.
- **Gradient Boosting Machines (GBM):** Utilizes GBM feature importance for feature selection.

#### 3. Stability Selection
- **Randomized Lasso/Logistic Regression:** Applies regularization over subsamples to select features.

#### 4. Recursive Feature Addition or Elimination
- **Recursive Feature Addition (RFA):** Begins with no features and adds the most improving ones.
- **Recursive Feature Elimination (RFE):** Starts with all features, removing the least important iteratively.

#### 5. Dimensionality Reduction Techniques
- **Principal Component Analysis (PCA):** Transforms features into principal components.
- **t-SNE:** Maps high-dimensional data to a lower-dimensional space.
- **Autoencoders:** Uses neural networks for unsupervised feature extraction.

#### 6. Genetic Algorithms
- **Evolutionary Feature Selection:** Employs evolutionary strategies to select optimal features.

#### 7. Feature Shuffling
- **Permutation Importance:** Assesses importance by the increase in prediction error when a feature's values are shuffled.

#### 8. Bayesian Model Averaging
- **Bayesian Approaches:** Incorporates model uncertainty into the feature selection process.

#### 9. Information-Theoretic Methods
- **Minimum Redundancy Maximum Relevance (mRMR):** Chooses features with high relevance and low redundancy.

#### 10. Biclustering
- **Biclustering Algorithms:** Finds clusters of features and samples for feature selection.